iterators.tex

```diff
@@ -623,31 +650,186 @@
+\range{i}{j} denotes a range. Let $N$ be the smallest number of
+applications of \tcode{++i} necessary to make \tcode{bool(i == j)} be
+\tcode{true}. \tcode{SizedIterator<I>()} is satisfied if and only if:
+
+\begin{itemize}
+\item If $N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{j - i} is well-defined and equals $N$.
+
+\item If $-N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{i - j} is well-defined and equals $-N$.
+\end{itemize}
+\end{itemdescr}
+
+\pnum
+The \tcode{disable_sized} type predicate provides a mechanism to enable
+use of iterator types that meet the syntactic requirements but do not
+in fact satisfy \tcode{SizedIterator} with the library.
 
 \pnum
-The \tcode{Sentinel} concept defines requirements for a type that
+\enternote A program that instantiates a library template that requires
+an \tcode{Iterator} with such an iterator type \tcode{I} is ill-formed
+with no diagnostic required unless \tcode{disable_sized<I>} evalutes to
+\tcode{true}~(\ref{structure.requirements}). \exitnote
+
+\enternote The \tcode{SizedIterator} concept is satisfied by
+\tcode{RandomAccessIterator}s~(\ref{iterators.random.access}) and by
+counted iterators~(\ref{counted.iterator}).\exitnote
+
+\ednote{This concept also gives us a way to demote the category of \tcode{move_iterator}s to Input
+while retaining the ability of \tcode{move_iterator} pairs to communicate the range's size to
+container constructors.}
+} %% color{newclr}
+
+\rSec2[iterators.sentinel]{Concept Sentinel}
+\pnum
+The \tcode{Sentinel} concept \oldtxt{defines requirements for a type that
 is an abstraction of the past-the-end iterator. Its values can be
-compared to an iterator for equality.
+compared to an iterator for equality.}\newtxt{specifies requirements on
+the types \tcode{S} and \tcode{I} of a sentinel and an iterator that
+denote a range.}
 
 \indexlibrary{\idxcode{Sentinel}}%
-\begin{codeblock}
-  template <class T, class I>
+\begin{itemdecl}
+  template <class @\oldtxt{T}\newtxt{S}@, class I>
   concept bool Sentinel() {
-    return Regular<T>() &&
+    return Regular<@\oldtxt{T}\newtxt{S}@>() &&
       Iterator<I>() &&
-      EqualityComparable<T, I>();
+      EqualityComparable<@\oldtxt{T}\newtxt{S}@, I>();
   }
-\end{codeblock}
+\end{itemdecl}
+
+{\color{newclr}
+\begin{itemdescr}
+Let \tcode{s} and \tcode{t} be values of type \tcode{S},
+and \tcode{i} a value of type \tcode{I} such that \range{i}{s}
+denotes a range. Types \tcode{S} and \tcode{I} satisfy
+\tcode{Sentinel<S, I>()} if and only if:
+
+\begin{itemize}
+\item \tcode{i == s}, \tcode{s == i},\tcode{i != s}, and
+      \tcode{s != i} are well-defined.
+
+\item \tcode{bool(i == s)} if and only if \tcode{i} and
+      \tcode{s} denote the same element.
+
+\item If \range{i}{s} and \range{i}{t} denote the same range,
+      then \tcode{s == t} is well-defined and \tcode{s}
+      equals \tcode{t}.
```

Comment:
> Transitivity of `==` between iterators and sentinels is the crux of the issue: if `i == s && j == s` implies `i == j`, then a sentinel must represent a position just like an iterator does.

That's exactly my point. `op==` for sentinels has to be consistent with the fact that sentinels represent *position in sequence*. It follows that comparing two sentinels can only make sense when they are denoting (logical) positions in the same sequence. Imagine this case:

```c++
template<Iterator I> struct counted_iterator {
  I it_;
  int count_;
  counted_iterator& operator++() {
    ++it_; ++count_; return *this;
  }
};

struct counted_sentinel {
  int count_;
};

bool operator==(counted_iterator<auto> i, counted_sentinel s) {
  return i.count_ == s.count_;
}

bool operator==(counted_sentinel s1, counted_sentinel s2) {
  return s1.count_ == s2.count_; // BUGBUG right or wrong?
}
```

Now imagine I do this:

```c++
int data[] = {1,2,3,4,5,6,7,8,9};
counted_iterator<int*> it0 { data + 1, 0 }, i1 { data + 1, 1 };
counted_sentinel s0 { 0 }, s1 { 1 };
```

First question: does `it0` equal `it1`? They have the same type and denote the same element ... but not in the same range. Next question: does `s0` equal `s1`? They have different state, but *could* denote the same element, depending on which iterator it's compared to.

It's easy to rearrange the example so that `s0` and `s1` have the *same* state but denote different elements depending on which iterator it's compared to.

IMO, the only logical conclusion is to say that the domain of `op==` for iterators and sentinels has to be limited to comparisons that happen within the same range. `op==` on sentinels has to respect it's semantics: it's a logical position in a range. Its `op==` is free to assume that -- since the user is comparing sentinels -- they must refer to the same range; otherwise, it's a domain violation. And since sentinels always refer to the last-plus-one element, it's always safe for sentinel `op==` to return true, ignoring the state of the sentinel.